### ISSUE TITLE
Issue with tf.keras.mixed_precision.experimental.LossScaleOptimizer

### ISSUE THREAD

**Issue by mattbernardini** [tensorflow/tensorflow#33484]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.0
- Python version: 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0, 7.5
- GPU model and memory: Titian RTX 24gb

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When attempting to use tf.keras.mixed_precision.experimental.LossScaleOptimizer it fails to cast a matmul to float16
**Describe the expected behavior**
Matmul should cast to float16
**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
import tensorflow as tf
import numpy as np
from tensorflow.keras.datasets import mnist

gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    try:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
        print(len(gpus), "Physical GPUS,", len(logical_gpus), "Logical GPUS")
    except RuntimeError as e:
        print(e)

tf.keras.backend.set_floatx('float16')
tf.keras.backend.set_epsilon(1e-4)

model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(tf.keras.layers.MaxPooling2D((2, 2), padding="same"))
model.add(tf.keras.layers.Conv2D(64, (3, 3), activation='relu', padding="same"))

model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(64, activation='relu'))
model.add(tf.keras.layers.Dense(10, activation='softmax'))

model.summary()

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype(np.float16) / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype(np.float16) / 255

train_labels = tf.keras.utils.to_categorical(train_labels, dtype='float16')
test_labels = tf.keras.utils.to_categorical(test_labels, dtype='float16')

opt = tf.keras.optimizers.RMSprop()
opt = tf.keras.mixed_precision.experimental.LossScaleOptimizer(opt, "dynamic")

model.compile(optimizer=opt,
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(tf.dtypes.cast(train_images, tf.float16), tf.dtypes.cast(train_labels, tf.float16), epochs=50, batch_size=64, steps_per_epoch=200)

test_loss, test_acc = model.evaluate(test_images, test_labels)

test_acc

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gadagashwini** [comment]
I could reproduce the issue with Tf 2.0. Please take a look at the colab [gist](https://colab.sandbox.google.com/gist/gadagashwini/7909678539c398aa1f4b8fafa334807f/untitled208.ipynb). Thanks!

**Comment by gowthamkpr** [comment]
@mattbernardini Try converting all the inputs to float32 and it should work fine. Please find my github gist [here](https://colab.sandbox.google.com/gist/gowthamkpr/2ab5ef2ed2af387abc2e60016a3936bc/untitled208.ipynb).

**Comment by mattbernardini** [comment]
Right, that works for float32, but does not work for float16 which was the original issue.

**Comment by gowthamkpr** [comment]
@reedwm Can you PTAL?

**Comment by RaphaelMeudec** [comment]
@gowthamkpr I'm running into the same issue, trying to run float16 training. Not sure why you're casting to float32 in the .fit() method? Thanks!

**Comment by verrannt** [comment]
I encounter the same problem with float16 training. So far I have figured out that the issue lies in the LossScaleOptimizer, because the issue disappears when you remove it (but then obviously losses are too tiny to be effective). 

To be more specific, the error is raised when the LossScaleOptimizer calls its `get_scaled_loss()` function which simply multiplies (via python operator *) the loss with the `loss_scale`, which is an instance of `tf.train.experimental.LossScale`. [See the code here.](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/mixed_precision/experimental/loss_scale_optimizer.py#L149)

```py
def get_scaled_loss(self, loss):
    loss_scale = self._loss_scale()
    if callable(loss):
      return lambda: loss() * loss_scale
    else:
      return loss * loss_scale
```

The error says that the `y` input of the multiplication is float32 and the `x` input is float16. Since the multiplication in `get_scaled_loss()` is done like this: `loss * loss_scale`, this means that `loss_scale` is a float32. A few lines above we see that `loss_scale = self._loss_scale()`, and `self._loss_scale` is an instance of `tf.train.experimental.LossScale`.

Now, if we look into the source code of `tf.train.experimental.LossScale` [here](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/training/experimental/loss_scale.py) we see that the `__call__()` method of e.g. the DynamicLossScale (which is the one @mattbernardini uses) returns `ops.convert_to_tensor(self._current_loss_scale)`. If we then check the initialization of `self._current_loss_scale` we see that it is initialized as float32:

```python
self._current_loss_scale = self._add_weight(
    name='current_loss_scale',
    dtype=dtypes.float32,
    initial_value=self._initial_loss_scale)
```

So this is why the TypeError is thrown when using LossScaleOptimizer in float16 training. I am not exactly sure why LossScales were implemented to only return float32 values. There are also no clear online guides on how to do float16 training in TF 2.0, so it would be nice if someone who knows the details of how Tensorflow implements the optimization process could provide information on proper float16 training, because evidently neither `tf.keras.backend.set_floatx('float16')` nor `    tf.keras.mixed_precision.experimental.set_policy('mixed_float16')` seem to work out-of-the-box.

**Comment by reedwm** [comment]
@verrannt, good investigation! In TF 2.1, this will be fixed. Alternatively, you can use the nightly builds with `pip install tf-nightly-gpu` which has the fix now. This has been fixed by casting the float32 loss scale to the dtype of the loss.

For reference, to do float16 training (typically referred to as mixed precision training), you **always** want to do the following

```
tf.keras.mixed_precision.experimental.set_policy('mixed_float16')  # Good
```

You **do not** want to set floatx to float16 with the following, unless you also set the policy to 'mixed_float16':

```
tf.keras.backend.set_floatx('float16')  # Bad!
```

The reason you want to set the policy instead of floatx is that the policy will keep variables in float32, which is necessary for numeric stability.

As for an online guide to do float16 training, I agree we need one and we will have one in the future. For now, [the Policy API documentation](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/experimental/Policy) has a brief guide on how to do mixed precision training
